File: app/src/main/java/com/onemoresecret/msg_fragment_plugins/MsgPluginTotp.java
package com.onemoresecret.msg_fragment_plugins;


import androidx.fragment.app.Fragment;

import com.onemoresecret.MessageFragment;
import com.onemoresecret.OutputFragment;
import com.onemoresecret.R;
import com.onemoresecret.TotpFragment;
import com.onemoresecret.crypto.OneTimePassword;

public class MsgPluginTotp extends MsgPluginEncryptedMessage {
    public MsgPluginTotp(MessageFragment messageFragment, byte[] messageData) {
        super(messageFragment, messageData);
        context.getMainExecutor().execute(() -> {
            var message = new String(messageData);
            var totpFragment = ((TotpFragment) messageView);
            totpFragment.init(new OneTimePassword(message), messageView, code -> {
                ((OutputFragment) outputView).setMessage(code, context.getString(R.string.one_time_password));
                totpFragment.setTotpText(Boolean.TRUE.equals(messageFragment.getHiddenState().getValue()) ? "●".repeat(code.length()) : code);
            });

            //observe hidden state
            messageFragment.getHiddenState().observe(messageView, hidden -> totpFragment.refresh());
        });
    }

    @Override
    public Fragment getMessageView() {
        if (messageView == null) messageView = new TotpFragment();
        return messageView;
    }
}

File: app/src/main/java/com/onemoresecret/msg_fragment_plugins/MsgPluginTotp.kt
package com.onemoresecret.msg_fragment_plugins

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.width
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import androidx.fragment.app.FragmentActivity
import com.onemoresecret.R
import com.onemoresecret.crypto.OneTimePassword
import com.onemoresecret.composable.TotpCodeView
import kotlinx.coroutines.delay
import kotlin.time.Duration.Companion.milliseconds

class MsgPluginTotp(
    activity: FragmentActivity,
    messageData: ByteArray,
    onNavigateBack: () -> Unit
) : MsgPluginEncryptedMessage(activity, messageData, onNavigateBack) {

    private val otp = OneTimePassword(String(messageData))

    @Composable
    override fun MessageView(hiddenState: Boolean) {
        var code by remember { mutableStateOf("") }
        var remaining by remember { mutableIntStateOf(0) }
        val title = activity.getString(R.string.one_time_password)

        LaunchedEffect(Unit) {
            while (true) {
                val state = otp.state
                remaining = (otp.period - state.secondsUntilNext).toInt()
                val newCode = otp.generateResponseCode(state.current)
                if (code != newCode) {
                    code = newCode
                    outputViewModel.setMessage(newCode, title)
                }
                delay(500.milliseconds)
            }
        }

        TotpCodeView(
            code = code,
            remaining = "...${remaining}s",
            hiddenState = hiddenState
        )
    }
}
